<a href="https://colab.research.google.com/github/theysee/sberbank_reboot/blob/master/%D0%9F%D0%BE%D1%82%D0%BE%D0%BA_3_%D0%9F%D1%80%D0%B0%D0%BA%D1%82%D0%B8%D0%BA%D0%B0_%D0%BA_%D0%B7%D0%B0%D0%BD%D1%8F%D1%82%D0%B8%D1%8E_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Практика



*   [Презентация](https://docs.google.com/presentation/d/1CE5hOVm2PBSDXA9sK7WtikFV8rUZlYwA9hc8vVh_tgY/edit?usp=sharing)
*   [master](https://colab.research.google.com/drive/1TeKcA6LIpFABRfBPWEz8VTWtZzRwDIhR)
*   [Дамп базы](https://drive.google.com/open?id=1RJx-1kh8az1OSQqnsDjCTWdzJBlDkjcB)

![схема базы](http://bdis.umeta.ru/db/db_course/labs/SQLpHTM/king.gif)


In [0]:
!pip install PyMySQL

     |████████████████████████████████| 51kB 1.6MB/s 


In [0]:
!pip install pandas

In [0]:
import pymysql
import pandas as pd
# доступны student-ы от 0 до 49

DB_HOST = '89.223.95.235'
DB_USER = 'student_0'
DB_USER_PASSWORD = 'student_0'
DB_NAME = 'sample_db'

In [0]:
conn = pymysql.connect(DB_HOST, DB_USER, DB_USER_PASSWORD, DB_NAME)
pd.read_sql_query("select * from customer", conn)

customer_id  ...                                           comments
0           100  ...  very friendly people to work with -- sales rep...
1           101  ...  rep called 5/8 about change in order - contact...
2           102  ...  company doing heavy promotion beginning 10/89....
3           103  ...      contact rep about new line of tennis rackets.
4           104  ...  customer with high market share (23%) due to a...
5           105  ...  tends to order large amounts of merchandise at...
6           106  ...  support intensive. orders small amounts (< 800...
7           107  ...  first sporting goods store geared exclusively ...
8           108  ...                                                   
9           201  ...  large general-purpose sports store with an aff...
10          202  ...               specializes in basketball equipment.
11          203  ...               follow up on the promotion proposal.
12          204  ...  large floorspace.  prefers maintaining large a...
13          205  ...   tremendous potential for an exclusive agreement.
14          206  ...              contact rep. about new product lines.
15          207  ...  customer requires written price quotes before ...
16          208  ...  very personal purchasing agents -- sharon and ...
17          211  ...  have an open purchase order for $3000.  ship i...
18          212  ...  pursue a contract -- possible candidate for vo...
19          213  ...                may be moving to a larger location.
20          214  ...                 target market is serious athletes.
21          215  ...  target market is casual and weekend athletes. ...
22          216  ...   customer prefers to be called between 10 and 12.
23          217  ...                      general purpose sports store.
24          218  ...  store does not open until 11am and does not ha...
25          221  ...          discount bicycle and sporting good store.
26          222  ...                          exclusive bicycle dealer.
27          223  ...      clerk answers all phone lines.  ask for mike.
28          224  ...  call joe to make sure last shipment was complete.
29          225  ...                     store catering to triathletes.
30          226  ...   customer in the midst of a cost-cutting program.
31          227  ...       customer referred to us by the all american.
32          228  ...  recently acquired another sporting goods store...

[33 rows x 11 columns]

### Задание 1
Выбрать коды всех городов, в которых расположены отделы фирмы.

In [0]:
pd.read_sql_query("select distinct location_id from department", conn)

location_id
0          122
1          123
2          124
3          167

### Задание 2
Для каждого сотрудника определить, какой процент по отношению к зарплате составляют его комиссионные.

In [0]:
pd.read_sql_query("""
SELECT employee_id,
       commission/salary * 100 AS 'percent'
FROM employee
WHERE commission IS NOT NULL
""", conn)

employee_id     percent
0         7499   18.750000
1         7521   40.000000
2         7555   24.000000
3         7557   96.000000
4         7564   24.000000
5         7600   72.000000
6         7654  112.000000
7         7789   66.666667
8         7820   61.538462
9         7844    0.000000

In [0]:
# более правильно
pd.read_sql_query("""
SELECT employee_id,
       COALESCE(commission,0)/salary * 100 AS 'percent'
FROM employee
ORDER BY percent DESC
""", conn)

employee_id     percent
0          7654  112.000000
1          7557   96.000000
2          7600   72.000000
3          7789   66.666667
4          7820   61.538462
5          7521   40.000000
6          7555   24.000000
7          7564   24.000000
8          7499   18.750000
9          7919    0.000000
10         7916    0.000000
11         7902    0.000000
12         7934    0.000000
13         7900    0.000000
14         7950    0.000000
15         7876    0.000000
16         7844    0.000000
17         7839    0.000000
18         7369    0.000000
19         7954    0.000000
20         7799    0.000000
21         7788    0.000000
22         7782    0.000000
23         7698    0.000000
24         7676    0.000000
25         7609    0.000000
26         7569    0.000000
27         7566    0.000000
28         7560    0.000000
29         7507    0.000000
30         7506    0.000000
31         7505    0.000000

### Задание 3
Вывести два инициала (с точками) и фамилии всех сотрудников, например:

In [0]:
pd.read_sql_query("""

  SELECT
    CONCAT(SUBSTR(first_name, 1, 1), '.') as I1,
    CONCAT(middle_initial, '.') as I2,
    last_name
  FROM
    employee

""", conn)

I1  I2 last_name
0   j.  q.     smith
1   k.  j.     allen
2   j.  k.     doyle
3   l.  s.    dennis
4   l.  d.     baker
5   c.  d.      ward
6   d.  t.    peters
7   k.  p.      shaw
8   s.  s.    duncan
9   g.  j.     lange
10  t.  m.     jones
11  c.  l.   alberts
12  r.  y.    porter
13  r.  m.     lewis
14  k.  j.    martin
15  d.  d.   sommers
16  m.  s.     blake
17  c.  f.     clark
18  d.  t.     scott
19  l.  n.      west
20  m.  g.    fisher
21  p.  s.      ross
22  f.  a.      king
23  m.  a.    turner
24  d.  g.     adams
25  f.  s.     james
26  j.  d.      ford
27  g.  m.   roberts
28  m.  a.   douglas
29  b.  m.    miller
30  a.  b.    jensen
31  j.  t.    murray

### Задание 4
По каждой сделке вывести точную сумму сделки, сумму сделки, округленную в большую сторону, округленную в меньшую сторону, округленную по общепринятым правилами округления.

### Задание 5
Выбрать фамилии всех сотрудников, у которых комиссионные больше зарплаты.

### Задание 6
Выбрать фамилии всех сотрудников, фамилии которых начинаются на букву 'S'.

### Задание 7
Выбрать имена всех сотрудников, которые не являются менеджерами (job_id=671) и не работают в отделе SALES в NEW YORK (department_id=13).

### Задание 8
Выбрать фамилии всех сотрудников, у которых код должности 670 или 677 (CLERK или SALESPERSON).

### Задание 9
Выбрать всех сотрудников, имена которых состоят из 6 букв и начинаются на 'MAR'.

### Задание 10
Выбрать фамилии всех сотрудников, которые поступили на работу после 15 апреля 1985 года.

### Задание 11
Выбрать фамилии всех сотрудников, которые поступили на работу в 1985 году.

### Задание 12
Для каждого сотрудника выбрать количество полных лет работы в фирме.

### Задание 13
Выбрать количество сотрудников, получающих комиссионные.

### Задание 14
Выбрать количество и общую сумму сделок, совершенных с покупателем, код которого - 104.

### Задание 15
Выбрать среднюю зарплату по каждой должности.

### Задание 16
Выбрать среднюю зарплата продавцов (код должности - 670).

### Задание 17
 Выбрать средние зарплаты продавцов (код должности - 670) и клерков (код должности - 667).

### Задание 18
Выбрать коды продуктов, по которым было совершено меньше 10 продаж.



### Задание 19
Выбрать максимальную зарплату продавцов (код должности - 670) по каждому отделу.

### Задание 20
Выбрать список сотрудников фирмы в алфавитном порядке.